### PR TITLE
minor refactoring of term list 

### PR DESCRIPTION
**Pull request by norcross**
displayed terms are now only ones that have been applied to the post itself. this addresses #7

### CODE REVIEW

File: includes/meta.php [hunk 1]
```diff
@@ -41,9 +41,23 @@ function sticky_post_for_term( $post_id, $term_id ) {
...
-	$taxonomies = array_keys( get_taxonomies( [
+
+	/**
+	 * Retrieve an array of post types that should be eligible for taxonomy-based stickiness.
+	 *
+	 * @param array $post_types Post types that should be able to use Sticky Tax.
+	 */
+	$post_types = apply_filters( 'stickytax_post_types', array( 'post' ) );
+
+	// Bail if we have no post types.
```
Comment (stevegrunwell) [inline includes/meta.php]: Good call. Might even make sense to drop this filter and instead register post type support for 'sticky-tax', then query post types using [`get_post_types_by_support()`](https://developer.wordpress.org/reference/functions/get_post_types_by_support/), but a filter works for now.

**Review by stevegrunwell — COMMENTED** [review]
I'm not sure if I agree with the explicit server-side filtering of the post term list. My hope was to add/remove terms from the list (and de-select them, when appropriate) without requiring a page refresh, which would be much more JavaScript-driven.

Scenarios:
1. When I'm creating a new post and publishing immediately I'd expect to be able to assign my new post to be sticky.
    * With this PR, it would require an explicit second click of the save button.
2. When editing a post, I'd expect to be able to re-assign the post to another existing taxonomy and make the term sticky in a single click of the save button.
    * This is currently possible and would continue to be in a JS-powered solution, but would not be possible if the list was filtered server-side.
3. If I create a new taxonomy term as I'm authoring a post, I'd expect to be able to assign the post to be sticky in that term right away.
    * Neither the current nor this approach address this.

Do you feel that presenting the entire, unfiltered list (e.g. #7) should be considered a blocker for the version 1.0.0 release, or is it something we can reasonably address in the next release?

Also, please ensure that tests are updated when plugin functionality is changed — [this PR currently has multiple failing unit tests](https://travis-ci.org/liquidweb/sticky-tax/builds/267350223).

**Comment by norcross** [comment]
I personally think showing all available terms is a bad idea for numerous reasons. 

1. it's a big ass list. my test had 5 taxonomies and over 100 items.

2. it allows you to select a term that the post is not applied to, which then does nothing. 

3. adding / refreshing the term list is a good idea, but that can be done regardless of whether it shows it all or not.

**Comment by stevegrunwell** [comment]
1. Part of the reason for Select2 (though mostly because many average users have a helluva time with `<select multiple>`) — would it make sense to instead use a filter to determine which taxonomies appear, rather than automatically loading anything with `public => true`?
2. No argument here, but doing nothing seems better than forcing multiple saves. Perhaps filtering out sticky term IDs on the `save_post` callback would make more sense (e.g. "oh, you assigned it to terms A, B, and C but only A and B are assigned? We won't save the C reference, then")?
3. That's a fair point, but case number 2 above is still removed by this PR.

**Comment by norcross** [comment]
I would have used checkboxes with assigned terms only. and hell, if you wanted to get wild, you could hook into the click action of adding a term and then add the new item markup

File: assets/css/sticky-tax.min.css [hunk 2]
```diff
@@ -0,0 +1 @@
+.term-sticky-list-wrap .term-sticky-list-group{margin-bottom:10px;padding-bottom:10px;border-bottom:1px dotted #ccc}.term-sticky-list-wrap .term-sticky-list-group:last-of-type{margin-bottom:0;padding-bottom:0;border-bottom:0 none}.term-sticky-list-wrap h4{margin:0 0 10px 0;font-size:1.1em}.term-sticky-list-wrap .term-sticky-list{margin:0}.term-sticky-list-wrap .term-sticky-list-empty{font-style:italic;margin:0}.term-sticky-list-wrap .term-sticky-list-empty.term-sticky-list-show{display:block}.term-sticky-list-wrap .term-sticky-list-empty.term-sticky-list-hide{display:none}
```
Comment (stevegrunwell) [inline assets/css/sticky-tax.min.css]: We should avoid committing the compiled assets to Git — they can live in SVN (out of necessity), but this file should be removed from the Git repo + blocked via the `.gitignore` file.

File: assets/js/sticky-tax.js [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+ * Scripting for the various message actions.
+ *
+ * @package LiquidWeb\StickyTax
+ * @author  Liquid Web
+ */
+
+/**
+ * Add our newly selected item to the box.
+ *
+ * @param  {[type]} termID   [description]
```
Comment (stevegrunwell) [inline assets/js/sticky-tax.js]: This documentation doesn't tell us anything we can't derive from the variable names — we should either remove the `@param` and `@return` tags or, preferably, complete them.

File: assets/js/sticky-tax.js [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+ * @param  {[type]} termName [description]
+ * @param  {[type]} termType [description]
+ * @return {[type]}          [description]
+ */
+function stickyTaxAddItem( termID, termName, termType ) {
```
Comment (stevegrunwell) [inline assets/js/sticky-tax.js]: This and `stickyTaxRemoveFunction()` are both being inserted into the global namespace for no particular reason — it would be safer to wrap the function definitions within an IIFE.

File: assets/js/sticky-tax.js [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+	// Trim my term name to be safe.
+	termName = jQuery.trim( termName );
+
+	// Set an empty var.
+	var newItem = '';
+
+	// Build my new list item.
```
Comment (stevegrunwell) [inline assets/js/sticky-tax.js:25]: This may be more paranoia-driven, but it's generally safer and more performant to explicitly construct the element rather than relying on concatenating markup, e.g.
```js
var list = document.getElementById('list-' + termType);
  newItem = document.createElement('li'),
  label = document.createElement('label'),
  input = document.createElement('input');

newItem.id = termID;
newItem.dataset.termId = termID;
newItem.dataset.termName = termName;
newItem.classList.add('term-sticky-list-item');

label.classList.add('list-item-label');
label.htmlFor = 'list-item-' + termId;

input.type = 'checkbox';
input.name = 'sticky-tax-term-id[]';
input.id = 'list-item-' + termID;
input.value = termID;

label.appendChild(input);
label.appendChild(document.createTextNode(termName));
newItem.appendChild(label);
list.appendChild(newItem);
```

File: assets/js/sticky-tax.js [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+function stickyTaxRemoveItem( termName, termType ) {
+
+	// Trim my term name to be safe.
+	termName = jQuery.trim( termName );
+
+	// Find the item in the term group with our term name.
+	jQuery( 'div#term-sticky-' + termType + '-group li[data-term-name="' + termName + '"]' ).filter( function( index ) {
+
+		// Uncheck the box to be safe.
+		jQuery( this ).find( 'input' ).prop( 'checked', false );
```
Comment (stevegrunwell) [inline assets/js/sticky-tax.js]: In scenarios like this, it's helpful to cache selectors like `jQuery( this )`:

```js
var $self = jQuery( this );

$self.find( 'input' ).prop( 'checked', false );
$self.remove();
```

Additionally, [`jQuery.remove()` will remove the entire element from the DOM](http://api.jquery.com/remove/), so there shouldn't be any reason to first explicitly un-check the checkbox.

File: assets/js/sticky-tax.js [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+		// Watch for any clicking inside that div.
+		$( '#tagsdiv-post_tag' ).on( 'click', function( event ) {
+
+			// Check for the add new button.
+			if ( $( event.target ).is( 'input.tagadd' ) ) {
+
+				// Get my term name.
+				termName = $( '.tagchecklist button:last' ).parent( 'span' ).first().contents().filter( function() {
+					return this.nodeType == 3;
```
Comment (stevegrunwell) [inline assets/js/sticky-tax.js]: When checking node types, it's generally preferable to [use the `Node.nodeType` constants](https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType#Constants) rather than referring to their values directly. These evaluations should also be done using a strict `===` comparator to avoid ambiguous results.

File: assets/js/sticky-tax.js [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+	if ( itemCount < 1 ) {
+		jQuery( 'div#term-sticky-' + termType + '-group' ).find( 'p.term-sticky-list-empty' ).removeClass( 'term-sticky-list-hide' ).addClass( 'term-sticky-list-show' );
+	}
+
+}
+
+/**
+ * Start the engines.
+ */
+jQuery( document ).ready( function($) {
```
Comment (stevegrunwell) [inline assets/js/sticky-tax.js:78]: This (and all other functions) should declare [ECMAScript strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode) — if the JS is moved within an IIFE, it can be set once.

File: assets/js/sticky-tax.min.js [hunk 9]
```diff
@@ -0,0 +1 @@
...roup").find("p.term-sticky-list-empty").removeClass("term-sticky-list-hide").addClass("term-sticky-list-show")}jQuery(document).ready(function(t){t.fn.divExists=function(t){var e=[].slice.call(arguments,1);return this.length&&t.call(this,e),this};var e,i,s;t("div#categorydiv").divExists(function(){t("ul.categorychecklist").on("change","input",function(s){e=t(this).attr("value"),i=t.trim(t(this).parent("label").text()),t(this).is(":checked")?stickyTaxAddItem(e,i,"category"):stickyTaxRemoveItem(i,"category")})}),t("div#tagsdiv-post_tag").divExists(function(){t("#tagsdiv-post_tag").on("click",function(e){if(t(e.target).is("input.tagadd")){i=t(".tagchecklist button:last").parent("span").first().contents().filter(function(){return 3==this.nodeType}).text(),i=t.trim(i),s=t("input#sticky-tax-nonce").val();var a={action:"stickytax_get_id_from_name",term_name:i,term_type:"post_tag",nonce:s};jQuery.post(ajaxurl,a,function(t){if(!0!==t.success)return!1;""!==t.data.term_id&&stickyTaxAddItem(t.data.term_id,i,"post_tag")})}t(e.target).is("button.ntdelbutton")&&stickyTaxRemoveItem(i=t(e.target).parent("span").first().contents().filter(function(){return 3==this.nodeType}).text(),"post_tag")})})});
```
Comment (stevegrunwell) [inline assets/js/sticky-tax.min.js]: Another file that we're better off adding to the `.gitignore`.

File: includes/ajax.php [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+			'errcode' => 'NO_TERM_TYPE',
+			'message' => __( 'No term type was provided.', 'sticky-tax' ),
+		);
+
+		// And handle my JSON return.
+		wp_send_json_error( $return );
+	}
+
+	// Set my variables.
+	$name   = preg_replace( '/\PL/u', '', sanitize_text_field( $_POST['term_name'] ) );
```
Comment (stevegrunwell) [inline includes/ajax.php]: I'm not entirely clear on what your `preg_replace()`is doing here.

File: includes/ajax.php [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+	$type   = trim( sanitize_text_field( $_POST['term_type'] ) );
```
Comment (stevegrunwell) [inline includes/ajax.php]: The `trim()` is unnecessary, as [`sanitize_text_field()` will already trim the output](https://developer.wordpress.org/reference/functions/sanitize_text_field/) (by way of `_sanitize_text_fields()`).

File: includes/meta.php [hunk 12]
```diff
@@ -38,12 +38,31 @@ function sticky_post_for_term( $post_id, $term_id ) {
...
 
 /**
- * Register the Sticky meta box on applicable post types.
+ * Fetch the taxonomies tied to a particular object.
+ *
+ * @param  object  $object  The object we're looking against.
+ * @param  string  $output  The format to output.
+ * @param  boolean $public  Whether to remove the non-public ones or not.
+ *
+ * @return array
```
Comment (stevegrunwell) [inline includes/meta.php:47]: Based on the `return` statement, this could also be a boolean `FALSE`.

File: includes/meta.php [hunk 13]
```diff
@@ -83,46 +125,97 @@ function register_meta_boxes() {
...
-				</optgroup>
-			<?php endforeach; ?>
-		</select>
-	</p>
+		// Set a class for the message.
+		$class  = ! empty( $data['items'] ) ? 'term-sticky-list-hide' : 'term-sticky-list-show';
 
-	<p class="description"><?php esc_html_e( '"Stick" this post to the top of the archive pages for the selected term(s)', 'sticky-tax' ); ?></p>
+		// Wrap each group in a div.
+		echo '<div id="term-sticky-' . esc_attr( $type ) . '-group" class="term-sticky-list-group" data-term-type="' . esc_attr( $type ) . '" >';
```
Comment (stevegrunwell) [inline includes/meta.php:171]: More a matter of personal preference, but I'm curious: how would you feel about reverting this back to a block of HTML embedded within the PHP, rather than a series of echo statements? I've found that in situations of >= 4 lines or so it can become much more readable, and IDEs gain the benefit of syntax highlighting on the markup.

For example:
```php
  // Setup work.
?>

  <h4><?php echo esc_html( $data['label'] ); ?></h4>
  <ul id="list-<?php echo esc_attr( $type ); ?>" class="term-sticky-list term-sticky-list-<?php echo esc_attr( $type ); ?>">
    <?php foreach ( $data['items'] as $term_id => $term_name ) : ?>

      <li id="item-<?php echo esc_attr( $term_id ); ?>" data-term-id="<?php echo esc_attr( $term_id ); ?> class="term-sticky-list-item">
        <!-- ... -->
      </li>

    <?php endforeach; ?>
  </ul>

<?php
  // Resume PHP-intensive stuff.
```

File: phpcs.xml [hunk 14]
```diff
@@ -35,6 +37,12 @@
 		<exclude-pattern>tests/*</exclude-pattern>
 	</rule>
 
-	<rule ref="WordPress-Extra" />
+	<rule ref="WordPress-Extra">
+		<properties>
+			<property name="custom_test_class_whitelist" type="array" value="WP_Ajax_UnitTestCase"/>
```
Comment (stevegrunwell) [inline phpcs.xml]: Did not know this was a configuration option within PHPCS.

Brent Rambo giving a thumbs up

File: tests/test-ajax.php [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+	public function test_ajax_success() {
+		$_POST['nonce']     = wp_create_nonce( 'sticky-tax' );
+		$_POST['action']    = 'stickytax_get_id_from_name';
+		$_POST['term_name'] = 'Uncategorized';
+		$_POST['term_type'] = 'category';
+
+		try {
+			$this->_handleAjax( 'stickytax_get_id_from_name' );
+		} catch ( WPAjaxDieContinueException $e ) {
+			return;
```
Comment (stevegrunwell) [inline tests/test-ajax.php]: Returning early from a test will consider it to have passed (since no assertions failed), which is likely not the intended result. Instead, I'd recommend dropping the `try/catch` block (thus letting any errors surface to the test runner) or explicitly failing the test if something goes wrong:

```php
try {
  // Something that might fail.

} catch ( MyException $e ) {
  $this->fail( 'Caught unexpected exception: $e->getMessage() );
}
```

File: tests/test-ajax.php [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+		}
+
+		// Check that the exception was thrown.
+		$response = json_decode( $this->_last_response );
+
+		$this->assertInternalType( 'object', $response );
```
Comment (stevegrunwell) [inline tests/test-ajax.php]: These assertions can be simplified somewhat — if `$response` isn't an object, `$this->assertObjectHasAttribute()` will throw a warning, which will fail the test. Same for `$response->data`.

Explicitly testing isn't a bad idea, it's just not 100% necessary when you have checks below that will implicitly assert `$response` is an object.

File: tests/test-ajax.php [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+		$this->assertObjectHasAttribute( 'success', $response );
+		$this->assertFalse( $response->success );
+		$this->assertObjectHasAttribute( 'data', $response );
+		$this->assertObjectHasAttribute( 'errcode', $response->data );
+		$this->assertEquals( $response->data->errcode, 'BAD_NONCE' );
+	}
+
+	public function test_ajax_error_no_term_name() {
```
Comment (stevegrunwell) [inline tests/test-ajax.php]: When you have multiple tests that follow the same sort of format, it can be helpful to create a separate method within the test suite to handle a lot of the common setup work. For example, you might add the following:

```php
/**
 * Send a request to the stickytax_get_id_from_name ajax handler.
 *
 * @param string $term_name The term name to submit.
 * @param string $term_type The taxonomy to which the term belongs.
 * @return object The JSON-decoded ajax response.
 */
protected function make_ajax_request( $term_name, $term_type ) {
  $_POST = [
    'nonce'     => wp_create_nonce( 'sticky-tax' ),
    'action'    => 'stickytax_get_id_from_name',
    'term_name' => $term_name,
    'term_type' => $term_type,
  ];

  $this->_handleAjax( 'stickytax_get_id_from_name' );

  return json_decode( $this->_last_response );
}
```

Each of the other tests would then become a matter of:

```php
public function test_ajax_error_no_term_name() {
  $response = $this->make_ajax_request( '', 'category' );

  $this->assertFalse( $response->success );
  $this->assertEquals( 'NO_TERM_NAME', $response->data->errcode );
}
```

This makes each test much clearer about what is actually being tested, and what is just setup.

File: tests/test-ajax.php [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+
+		// Check that the exception was thrown.
+		$response = json_decode( $this->_last_response );
+
+		$this->assertInternalType( 'object', $response );
+		$this->assertObjectHasAttribute( 'success', $response );
+		$this->assertFalse( $response->success );
+		$this->assertObjectHasAttribute( 'data', $response );
+		$this->assertObjectHasAttribute( 'errcode', $response->data );
+		$this->assertEquals( $response->data->errcode, 'NO_TERM_NAME' );
```
Comment (stevegrunwell) [inline tests/test-ajax.php]: When using PHPUnit's `assertEquals()` method, the _expected_ value should be the first parameter, while the _actual_ value comes second.

File: tests/test-ajax.php [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+		$this->assertFalse( $response->success );
+		$this->assertObjectHasAttribute( 'data', $response );
+		$this->assertObjectHasAttribute( 'errcode', $response->data );
+		$this->assertEquals( $response->data->errcode, 'NO_TERM_TYPE' );
+	}
+
+	public function test_ajax_error_no_term_found() {
+		$_POST['nonce']     = wp_create_nonce( 'sticky-tax' );
+		$_POST['action']    = 'stickytax_get_id_from_name';
+		$_POST['term_name'] = uniqid();
```
Comment (stevegrunwell) [inline tests/test-ajax.php]: I'm a huge fan of `uniqid()` in tests and use it regularly, but it may not be the best representation of a term name (especially since WordPress can be kinda messy in how it handles finding terms by names). Instead, I'd recommend using a hard-coded string, but adding a pre-condition assertion:

```php
$_POST['term_name'] = 'this-term-does-not-exist';

// Verify that the non-existent term is, in fact, non-existent.
$this->assertFalse(
  (bool) term_exists( $_POST['term_name'], 'category',
  'Category "' . $_POST['term_name'] . '" expected not to exist.'
);
```

File: tests/test-meta.php [hunk 20]
```diff
@@ -325,28 +317,21 @@ public function register_script_hooks_provider() {
 	}
 
 	public function test_register_scripts_checks_for_existing_scripts() {
```
Comment (stevegrunwell) [inline tests/test-meta.php:319]: We can safely drop some of the tests that were playing the "oh, if someone else has already registered Select2, we can use their version" — we should be the only ones registering/enqueueing Sticky Tax scripts + styles.

**Review by stevegrunwell — CHANGES_REQUESTED** [review]
I'm glad we had our call today so we could talk through some of these items. I've added some suggestions re: tests and noted a few items that could be problematic, and we'll work on refactoring the JS together once we have all of the use-cases within WP admin identified and addressed.

File: tests/test-ajax.php (hunk 15 above)
Comment (norcross) [reply tests/test-ajax.php]: I initially had a try/catch in there, but that's now how the Ajax unit tests are supposed to run based on all the documentation I found

File: includes/meta.php (hunk 13 above)
Comment (norcross) [reply includes/meta.php:171]: it's personal preference. I prefer the better syntax highlighting on the PHP functions that happens when echo'd, and it reads more disjointed to me when individual tags are echo'd inside of markup, but this ain't a hill worth dying on.

File: includes/ajax.php (hunk 10 above)
Comment (norcross) [reply includes/ajax.php]: it's tied to how I have to pull in the tag name. the JS sometimes adds in weird invisible line spacing and whatnot that isn't removed from just `sanitize_text_field` for some reason. this is one of those "regardless of what you throw at me, I will handle it" functions.

File: includes/meta.php (hunk 12 above)
Comment (norcross) [reply includes/meta.php:47]: agreed. will update.

File: includes/ajax.php (hunk 11 above)
Comment (norcross) [reply includes/ajax.php]: see https://github.com/liquidweb/sticky-tax/pull/8#discussion_r135582684

File: tests/test-ajax.php (hunk 15 above)
Comment (norcross) [reply tests/test-ajax.php]: to clarify: if you can get that working as expected with a try / catch, I'm all for it.

Comment (norcross) [reply tests/test-ajax.php]: to clarify: I tried a few different ways of structuring that test, this is the only one that seemed to work correctly.

File: tests/test-ajax.php (hunk 17 above)
Comment (norcross) [reply tests/test-ajax.php]: can you confirm this works in testing? When I do it, it fails on confirming object type.

Comment (stevegrunwell) [reply tests/test-ajax.php]: I've pushed an update to this branch that simplifies the tests in two ways:

1. Introduce the `make_ajax_request()` method, which accepts an array of arguments to merge into some defaults (that way, the test case only needs to specify which overrides to make).
2. Further simplify the `test_ajax_error_*` methods by using a [PHPUnit data provider](https://phpunit.de/manual/current/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers), so we only have to write the test logic once.

File: tests/test-ajax.php (hunk 15 above)
Comment (stevegrunwell) [reply tests/test-ajax.php]: It turns out you're right — because [`wp_send_json()` (and its related functions) use `wp_die()` when handling ajax requests](https://developer.wordpress.org/reference/functions/wp_send_json/), our tests have to explicitly catch the exceptions. Which is...well, one of the many reasons this isn't technically *unit* testing.

Instead, we want to use a `try/catch` but not return from within the `catch` portion, as that would bypass our test assertions, potentially leading to false positives. I've taken care of this in 83ed2c6e2.

File: includes/ajax.php (hunk 10 above)
Comment (stevegrunwell) [reply includes/ajax.php]: Huh, I learned something today. Apparently `\pL` (lowercase, not uppercase) is short-hand for `\p{L}`, [which matches "any kind of letter from any language."](http://www.regular-expressions.info/unicode.html). A comment to this effect (or perhaps an abstracted `better_sanitize_text_field()` function) would be helpful.

File: tests/test-meta.php (hunk 20 above)
Comment (norcross) [reply tests/test-meta.php:319]: I removed all the Select2 stuff, just kept the name of the test function itself.

File: assets/js/sticky-tax.js (hunk 4 above)
Comment (norcross) [reply assets/js/sticky-tax.js]: I'm not 100% sure what that means.

File: assets/js/sticky-tax.js (hunk 6 above)
Comment (norcross) [reply assets/js/sticky-tax.js]: agreed re: caching selectors. will update.

File: assets/js/sticky-tax.js (hunk 4 above)
Comment (stevegrunwell) [reply assets/js/sticky-tax.js]: I suppose I misspoke — it's not the global namespace, so much as the global **scope**. Custom JavaScript variables and functions, especially in environments where they'll be interacting with other code, should be isolated from the global variable scope as much as possible.

[The 10up Engineering Best Practices covers this well and in more detail](https://10up.github.io/Engineering-Best-Practices/javascript/#dont-pollute-the-window-object), but the general idea is to protect our script from outside interference as much as possible, and only expose things into the global variable scope (the `window` object) when we need to.

File: tests/test-meta.php (hunk 20 above)
Comment (norcross) [reply tests/test-meta.php:319]: fwiw the test itself is still valid, even if it's an edge case now that we have our own CSS and JS

File: assets/js/sticky-tax.js (hunk 4 above)
Comment (norcross) [reply assets/js/sticky-tax.js]: I understand now (as an idea). I assume you're addressing this in the rest of the JS you're dealing with.
